unified-prototype/unified-plugin/plugin-android/src/main/java/org/gradle/api/experimental/android/library/StandaloneAndroidLibraryPlugin.java

```diff
@@ -38,6 +41,7 @@ public void apply(Project project) {
         super.apply(project);
 
         AndroidLibrary dslModel = getAndroidLibrary();
+        project.getExtensions().add(ANDROID_LIBRARY, dslModel);
```

Comment:
💭 It would be nice if there was a way to say `autoRegisterExtension = true` in the annotation itself, and then applying the plugin automatically registered it as part of `apply`.  Re-typing these lines every time is silly boilerplate.

Reply:
I'm not 100% convinced that removing the implicit registration is the best approach, but I think we had an array of bad choices and so we just picked the one with the fewest moving parts.  We needed an approach that would not require the model object to be available before the plugin was applied. 

One option we considered was the inversion of what you suggest, i.e. an option to say "don't automatically register this model as an extension".  This would allow us to keep the implicit registration, but disable it for software types that want to construct the model object themselves during the apply method.  The cost of this is extra public API and potentially confusing behavior if the user didn't set this in the annotation (i.e. Gradle expects the model object to be available, but it isn't).

Another option we considered was to move the implicit registration until _after_ the plugin was applied.  The issue with this is that you have to handle the situation where the plugin _does_ register the extension, or other error cases like it registers a different extension with the same name, or it registers it with a different, but compatible type, etc.

What we settled on was just punting on the implicit registration (which is nice, but not technically necessary) and removing any assumptions in Gradle that it would be registered.  In the absence of a clear "best" choice, removing something unnecessary seemed like the best option to quickly unblock the investigation that spawned this.  And at least there are consistent expectations, regardless of how you write your software type plugin.

Another option that just occurred to me would be to add an `add(class<T> publicType, String name, Provider<T> provider)` method to `ExtensionContainer`.  This would allow Gradle to implicitly register it early, but the object doesn't have to actually be available until something tries to retrieve it later.  Then, I think the only danger would be that the plugin itself might try to retrieve it before it's ready, but that's totally on the plugin author.  

A variation of above would be to expose the model object in a different way (not as an extension).  The primary motivation for exposing this as an extension is so that non-declarative scripts can configure the model in a way similar to declarative scripts.  But, we could expose these in non-declarative scripts in some other way (i.e. not through the extension container) that allow us to register it lazily early and then access it after the plugin is applied (including from non-declarative scripts).  This would obviously require some work to expose this new thing in non-declarative files.

In any event, I'm prepared to be convinced that some other solution is better, if you feel strongly that one of these (or another option) is what we should do instead.

Reply:
I wasn't thinking about implicit behavior at all.  It sounds like you thought that through and found reasons not to do it (at least not yet).  I was just thinking of adding to this an explicit way to opt-in to registration by asking for this in the annotation itself, in order to avoid some repetition in the body of `apply`.